src/diffusers/models/transformers/transformer_2d.py

```diff
@@ -153,94 +149,178 @@ def __init__(
-            else:
-                self.proj_in = conv_cls(in_channels, inner_dim, kernel_size=1, stride=1, padding=0)
-        elif self.is_input_vectorized:
-            assert sample_size is not None, "Transformer2DModel over discrete input must provide sample_size"
-            assert num_vector_embeds is not None, "Transformer2DModel over discrete input must provide num_embed"
-
-            self.height = sample_size
-            self.width = sample_size
-            self.num_vector_embeds = num_vector_embeds
-            self.num_latent_pixels = self.height * self.width
-
-            self.latent_image_embedding = ImagePositionalEmbeddings(
-                num_embed=num_vector_embeds, embed_dim=inner_dim, height=self.height, width=self.width
+        # 2. Initialize the right blocks.
+        if self.is_input_continuous:
+            self._init_continuous_input(
+                in_channels=in_channels, use_linear_projection=use_linear_projection, norm_type=norm_type
             )
+        elif self.is_input_vectorized:
+            self._init_vectorized_inputs(sample_size=sample_size, norm_type=norm_type)
         elif self.is_input_patches:
-            assert sample_size is not None, "Transformer2DModel over patched input must provide sample_size"
-
-            self.height = sample_size
-            self.width = sample_size
-
-            self.patch_size = patch_size
-            interpolation_scale = (
-                interpolation_scale if interpolation_scale is not None else max(self.config.sample_size // 64, 1)
-            )
-            self.pos_embed = PatchEmbed(
-                height=sample_size,
-                width=sample_size,
-                patch_size=patch_size,
+            self._init_patched_inputs(
                 in_channels=in_channels,
-                embed_dim=inner_dim,
+                sample_size=sample_size,
+                norm_type=norm_type,
+                patch_size=patch_size,
                 interpolation_scale=interpolation_scale,
+                caption_channels=caption_channels,
             )
 
-        # 3. Define transformers blocks
+        # Others.
+        self.out_channels = in_channels if out_channels is None else out_channels
+        self.gradient_checkpointing = False
+
+    def _init_continuous_input(self, in_channels, use_linear_projection, norm_type):
+        self.use_linear_projection = use_linear_projection
+        inner_dim = self.config.num_attention_heads * self.config.attention_head_dim
+
+        self.norm = torch.nn.GroupNorm(
+            num_groups=self.config.norm_num_groups, num_channels=in_channels, eps=1e-6, affine=True
+        )
+        if use_linear_projection:
+            self.proj_in = torch.nn.Linear(in_channels, inner_dim)
+        else:
+            self.proj_in = torch.nn.Conv2d(in_channels, inner_dim, kernel_size=1, stride=1, padding=0)
+
         self.transformer_blocks = nn.ModuleList(
```

Comment:
I don't think we need to define the transformer blocks inside each init method? Ideally the init methods should only configure attributes.

Reply:
actually, these `init*` methods are not configuring attributes at all  - they mostly initialize the modules. I think in that case, it makes sense to define transformer blocks inside each in that case, even though it is a common model, the code will be duplicated 

e.g. this is 

```python
def _init_continuous_input(self):
        self.norm = torch.nn.GroupNorm(
            num_groups=self.config.norm_num_groups, num_channels=self.in_channels, eps=1e-6, affine=True
        )
        if self.use_linear_projection:
            self.proj_in = torch.nn.Linear(self.in_channels, self.inner_dim)
        else:
            self.proj_in = torch.nn.Conv2d(self.in_channels, self.inner_dim, kernel_size=1, stride=1, padding=0)

        # Use out_channels or default to in_channels
        if self.use_linear_projection:
            self.proj_out = torch.nn.Linear(self.inner_dim, self.out_channels)
        else:
            self.proj_out = torch.nn.Conv2d(self.inner_dim, self.out_channels, kernel_size=1, stride=1, padding=0)
```

Reply:
But the arguments to the transformer blocks are all the same right? None of these modules being created in the init methods are being passed to the transformer blocks? Why duplicate?